magnum-cats-effect/src/main/scala/com/augustnagro/magnum/magcats/Transactor.scala

```diff
@@ -0,0 +1,158 @@
+package com.augustnagro.magnum.magcats
+
+import cats.*
+import cats.effect.kernel.*
+import cats.effect.kernel.Outcome.*
+import cats.effect.std.*
+import cats.effect.syntax.all.*
+import cats.syntax.all.*
+import com.augustnagro.magnum.*
+
+import java.sql.Connection
+import javax.sql.DataSource
+import scala.util.control.NonFatal
+
+class Transactor[F[_]: Sync] private (
+    dataSource: DataSource,
+    sqlLogger: SqlLogger,
+    connectionConfig: Connection => Unit,
+    semaphore: Option[Semaphore[F]]
+):
+  def withSqlLogger(sqlLogger: SqlLogger): Transactor[F] =
+    new Transactor(
+      dataSource,
+      sqlLogger,
+      connectionConfig,
+      semaphore
+    )
+
+  def withConnectionConfig(
+      connectionConfig: Connection => Unit
+  ): Transactor[F] =
+    new Transactor(
+      dataSource,
+      sqlLogger,
+      connectionConfig,
+      semaphore
+    )
+
+  def connect[A](f: DbCon ?=> A): F[A] =
+    val io =
+      Resource
+        .make(acquireConnection)(releaseConnection)
+        .use: cn =>
+          Sync[F].delay(connectionConfig(cn)) >>
+            Sync[F].interruptible(f(using DbCon(cn, sqlLogger)))
+
+    semaphore.fold(io)(sem => sem.permit.use { _ => io })
+
+  def transact[A](f: DbTx ?=> A): F[A] =
+    val io =
+      Resource
+        .make(acquireConnection)(releaseConnection)
```

Comment:
Since you repeat this a lot, you could save it in a variable; e.g.

```scala
val makeConnection =
 Resource.make(acquireConnection)(releaseConnection)
```